Make sure code_typet always has parameters property

peterschrammel opened the pull request:
The existence of the parameters property may depend on
whether parameters() has been called or not, which can
lead to unexpectedly unequal instructions in goto programs.

Fixes #2083

tautschnig reviewed (approved):
Just a procedural suggestion: split this in two commits, with the first one doing the changes to std_types.h and the second one being a clean revert of 0b3170d425.

kroening commented:
I would then do the same with return_type().
Also, I'd mark the () constructor as deprecated, as this is another case of "problem caused by partially-constructed object".

kroening reviewed (changes requested):
do same with return_type, and make () constructor deprecated.

peterschrammel committed:
Revert 0b3170d: Stabilize clinit wrapper function type parameters

smowton reviewed (approved):
As per @tautschnig, once @kroening's reservations are addressed

peterschrammel committed:
Make sure code_typet is fully constructed

The existence of the parameters and return_type properties may
depend on whether parameters() or return_type() has been called
or not, which can lead to unexpectedly unequal instructions
in goto programs.

peterschrammel committed:
Add two-param code_typet constructors and deprecate the default one

peterschrammel committed:
Use the two-param code_typet constructor

peterschrammel committed:
Test for incomplete code_typet construction

peterschrammel committed:
Add missing module definitions

peterschrammel commented:
@kroening, I've introduced new constructors, deprecated the old one and cleaned up its occurrences.

kroening commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -760,15 +760,14 @@ void java_bytecode_convert_classt::add_array_types(symbol_tablet &symbol_table)
+    code_typet clone_type(
+      {this_param},
+      java_reference_type(symbol_typet("java::java.lang.Object")));
 
```
Note that we've got java_lang_object_type()

tautschnig commented:
Not there isn't enough work in here already, but "Use the two-param code_typet constructor" would now also enable a lot of use of `const` for almost all of the `code_typet` objects being declared.

kroening commented:
Note that this also tweaks the remove_returns behavior, which may not have been intended.

tautschnig commented:
> Note that this also tweaks the remove_returns behavior, which may not have been intended.

In what way is that impacted here?

peterschrammel commented:
> Note that this also tweaks the remove_returns behavior, which may not have been intended.

~I deliberately changed that because silently ignoring the failure and returning a nil-irept through a code_typet (!) didn't seem appropriate to me. That case can only occur when the symbol table is out-of-sync with the goto-program, which is clearly a programming error elsewhere.~
Since this is changing behaviour, I will put that change into a separate commit.

peterschrammel committed:
Return original type instead of nil in original_return_type

which is unchanged when there is no #return_value variable

peterschrammel commented:
I revise my statement above. The code is actually used in cbmc/symex_coverage. However, I think that the call there can be simplified if `original_return_type` behaved slightly differently. @tautschnig, please check b29b043 whether this follows your intentions.

tautschnig commented:
> please check b29b043 whether this follows your intentions.

This looks good to me, and certainly improves the clarity of the code.

chrisr-diffblue reviewed (approved):
Looks OK to me.

tautschnig commented:
I don't think this should have been merged just yet. I have created #2287 to track this.
